libnavui-lifecycle/src/main/java/com/mapbox/navigation/ui/lifecycle/MapboxNavigationApp.kt

```diff
@@ -0,0 +1,135 @@
+package com.mapbox.navigation.ui.lifecycle
+
+import android.app.Application
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.LifecycleOwner
+import com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
+import com.mapbox.navigation.base.options.NavigationOptions
+import com.mapbox.navigation.core.MapboxNavigation
+import com.mapbox.navigation.ui.lifecycle.internal.CarAppLifecycleOwner
+import com.mapbox.navigation.ui.lifecycle.internal.MapboxNavigationOwner
+
+/**
+ * Manages a default lifecycle for [MapboxNavigation].
+ *
+ * Call [MapboxNavigationApp.setup] from your application to specify your [NavigationOptions].
+ * At any point in the application you can register or unregister observers for [MapboxNavigation].
+ *
+ * Attach any and all activities to your application to share an instance of [MapboxNavigation].
+ * ```
+ * class MyApplication : Application() {
+ *   override fun onCreate() {
+ *     MapboxNavigationApp.setup(this)
+ *       .attachAllActivities()
+ *       .registerObserver(locationObserver)
+ *   }
+ * }
+ * ```
+ *
+ * Register and unregister individual LifecycleOwners. Below is an example of creating a single
+ * activity that uses [MapboxNavigation].
+ *
+ * ```
+ * class MyActivity : ComponentActivity() {
+ *   override fun onCreate() {
+ *     MapboxNavigationApp
+ *       .setup(this)
+ *       .attach(this)
+ *   }
+ *   override fun onResume() {
+ *     MapboxNavigationApp
+ *       .registerObserver(locationObserver)
+ *   }
+ *   override fun onPause() {
+ *     MapboxNavigationApp
+ *       .unregisterObserver(locationObserver)
+ *   }
+ * }
```

Comment:
This approach currently does not work because just calling `attach` does not register the `mapboxNavigationOwner.carAppLifecycleObserver` into `carAppLifecycleOwner` and `MapboxNavigation` is never delivered. This only happens in `attachAllActivities`.

Reply:
> This approach currently does not work because just calling attach does not register the mapboxNavigationOwner.carAppLifecycleObserver into carAppLifecycleOwner and MapboxNavigation is never delivered. 

What do you mean? `MapboxNavigationApp.attach` -> `CarAppLifecycleOwner.attach` ->  `startedReferenceCounter`

And the `startedReferenceCounter` is responsible for counting the "started" lifecycle events.

```
    fun attach(lifecycleOwner: LifecycleOwner) = apply {
        carAppLifecycleOwner.attach(lifecycleOwner)
    }

CarAppLifecycleOwner {
    fun attach(lifecycleOwner: LifecycleOwner) {
        logI(TAG, Message("attach"))
        lifecycleOwner.lifecycle.addObserver(startedReferenceCounter)
    }
```

Reply:
Problem is that `mapboxNavigationOwner.carAppLifecycleObserver` is never registered with `carAppLifecycleOwner` when calling just `attach`. So even though `carAppLifecycleOwner` is started, the `mapboxNavigationOwner` is never notified and the instance of `MapboxNavigation` is not created.

Reply:
This should be fixed. Any behavior issues we see can be shown in the `MapboxNavigationAppDelegateTest`